### ISSUE TITLE
Incorrect error in extended interface when base interface has `this` member

### ISSUE THREAD

**Issue by RyanCavanaugh** [Microsoft/TypeScript#7276]
**TypeScript Version:** 

master

**Code**

```ts
interface Foo<T extends Bar> { x: T }

interface Bar {
	thing: this;
}

interface Something extends Bar {
	other: Foo<this>;
}
```

**Expected behavior:**
No error

**Actual behavior:** 
Error
> ```
error TS2344: Type 'this' does not satisfy the constraint 'Bar'.
  Type 'Something' is not assignable to type 'Bar'.
    Types of property 'thing' are incompatible.
      Type 'Something' is not assignable to type 'this'.
```
